File: src/main/java/co/bitshifted/xapps/backstage/deploy/ToolsRunner.java
/*
 * Copyright (c) 2020. Bitshift (http://bitshifted.co)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package co.bitshifted.xapps.backstage.deploy;

import co.bitshifted.xapps.backstage.exception.DeploymentException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import java.io.*;
import java.lang.module.ModuleFinder;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.util.spi.ToolProvider;

/**
 * @author Vladimir Djurovic
 */
@Component
@Slf4j
public class ToolsRunner {

	private ToolProvider jdeps;
	private ToolProvider jlink;

	@PostConstruct
	public void init() {
		jdeps = ToolProvider.findFirst("jdeps").orElseThrow();
		jlink = ToolProvider.findFirst("jlink").orElseThrow();
	}

	public Set<String> getApplicationModules(Path deploymentPackageDir) throws DeploymentException {
		Path mainArtifactPath;
		String mainArtifactModuleName;
		try {
			var xmlProcessor = new XmlProcessor(deploymentPackageDir.resolve("ignite-config.xml"));
			var mainArtifact = xmlProcessor.findMainArtifact();
			log.debug("Found main artifact: {}", mainArtifact);
			mainArtifactPath = deploymentPackageDir.resolve(mainArtifact.get("path"));
			mainArtifactModuleName = mainArtifactModuleName(mainArtifactPath);
		} catch (Exception ex) {
			log.error("Failed to parse XML configuration", ex);
			throw new DeploymentException(ex);
		}
		var outString = new StringWriter();
		var out = new PrintWriter(outString);
		var errString = new StringWriter();
		var err = new PrintWriter(errString);

		var argsList = new ArrayList<String>();
		argsList.add("--module-path");
		argsList.add(deploymentPackageDir.resolve("modules").toAbsolutePath().toString());
		argsList.add("--list-deps");
		argsList.add(mainArtifactPath.toAbsolutePath().toString());
		// if classpath dir exists, add it to arguments
		var classpath = deploymentPackageDir.resolve("classpath");
		if(Files.exists(classpath)) {
			log.debug("Classpath directory found, adding it to jdeps command");
			argsList.add("classpath/*");
		}
		log.debug("jdeps arguments: {}", argsList);
		var result = jdeps.run(out, err, argsList.toArray(new String[argsList.size()]));
		var modules = new HashSet<String>();
		if(result == 0){
			log.debug("jdeps output: {}", outString.toString());
			try(var reader = new BufferedReader(new StringReader(outString.toString()))){
				String line;
				while((line = reader.readLine()) != null) {
					modules.add(line.trim());
				}
			} catch(IOException ex) {
				log.error("Failed to read jdeps output", ex);
				throw new DeploymentException(ex);
			}

		} else { 
			log.error("Error running jdeps: {}", errString.toString());
			throw new DeploymentException(errString.toString());
		}
		if(mainArtifactModuleName != null) {
			modules.add(mainArtifactModuleName);
		}
		log.debug("Found modules: {}", modules);
		return modules;
	}

	private String mainArtifactModuleName(Path artifactPath) {
		var finder = ModuleFinder.of(artifactPath);
		var modules = finder.findAll();
		if(modules.isEmpty()) {
			return null;
		}
		return modules.iterator().next().descriptor().name();
	}

	public void createRuntimeImage(Set<String> modules, Path modulesDir, Path outputDir) throws DeploymentException {
		var argList = new ArrayList<String>();
		argList.add("--module-path");
		argList.add(modulesDir.toAbsolutePath().toString());
		argList.add("--add-modules");
		argList.add(createModulesArgs(modules));
		argList.add("--output");
		argList.add(outputDir.toAbsolutePath().toString());
		log.debug("jlink args: {}", argList);

		var outString = new StringWriter();
		var out = new PrintWriter(outString);
		var errString = new StringWriter();
		var err = new PrintWriter(errString);
		var result = jlink.run(out, err, argList.toArray(new String[argList.size()]));
		if(result == 0){
			log.debug("jlink output: {}", outString.toString());
		} else {
			log.error("Error running jlink: {}", errString.toString());
			throw new DeploymentException(errString.toString());
		}
	}

	private String createModulesArgs(Set<String> modules) {
		StringBuilder sb = new StringBuilder();
		modules.forEach(m -> sb.append(",").append(m));
		return sb.toString().substring(1);
	}
}

File: src/main/kotlin/co/bitshifted/backstage/service/deployment/builders/ToolsRunner.kt
/*
 *
 *  * Copyright (c) 2022  Bitshift D.O.O (http://bitshifted.co)
 *  *
 *  * This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 */

package co.bitshifted.backstage.service.deployment.builders

import co.bitshifted.backstage.BackstageConstants
import co.bitshifted.backstage.exception.BackstageException
import co.bitshifted.backstage.exception.DeploymentException
import co.bitshifted.backstage.exception.ErrorInfo
import co.bitshifted.backstage.util.logger
import java.io.BufferedReader
import java.io.PrintWriter
import java.io.StringReader
import java.io.StringWriter
import java.nio.file.Files
import java.nio.file.Path
import java.util.spi.ToolProvider
import kotlin.io.path.name

class ToolsRunner(val buildDir: Path) {

    private val logger = logger(this)
    private val jdeps = ToolProvider.findFirst("jdeps").orElseThrow()
    private val jlink = ToolProvider.findFirst("jlink").orElseThrow()

    fun getJdkModules(): Set<String> {
        val outString = StringWriter();
        val out = PrintWriter(outString);
        val errString = StringWriter();
        val err = PrintWriter(errString);

        val argsList = mutableListOf<String>()
        argsList.add("--module-path")
        argsList.add(buildDir.resolve(BackstageConstants.OUTPUT_MODULES_DIR).toFile().absolutePath)
        argsList.add("--list-deps")
        argsList.addAll(getAllJarsInDirectory(buildDir.resolve(BackstageConstants.OUTPUT_MODULES_DIR)))
        argsList.addAll(getAllJarsInDirectory(buildDir.resolve(BackstageConstants.OUTPUT_CLASSPATH_DIR)))
        logger.debug("jdeps arguments list: {}", argsList)
        val result = jdeps.run(out, err, *argsList.toTypedArray())
        val modules = mutableSetOf<String>()
        if (result == 0) {
            logger.debug("jdeps output: {}", outString.toString());
            BufferedReader(StringReader(outString.toString())).useLines {
                it.filter { line -> isJdkModule(line.trim()) }.forEach { line -> modules.add(line.trim()) }
            }
        } else {
            logger.error("Error running jdeps: {}", errString.toString())
            throw DeploymentException("Error running jdeps: $errString" )
        }
        logger.debug("Found JDK modules: {}", modules)
        return modules
    }

    fun createRuntimeImage(modules : Set<String>, moduleDirs : List<Path>, outputDir : Path) {
        logger.info("Creating runtime image...")
        val outString = StringWriter();
        val out = PrintWriter(outString);
        val errString = StringWriter();
        val err = PrintWriter(errString);

        val argsList = mutableListOf<String>()
        argsList.add("--verbose")
        argsList.add("--module-path")
        argsList.add(concatenateModulesPath(moduleDirs))
        argsList.add("--add-modules")
        argsList.add(createModulesArgs(modules))
        argsList.add("--output")
        argsList.add(outputDir.toAbsolutePath().toString())
        logger.debug("jlink arguments: {}", argsList)

        val result = jlink.run(out, err, *argsList.toTypedArray())
        if (result == 0) {
            logger.info("Runtime image generated successfully")
        } else {
            logger.error("Error running jlink: {}", errString)
            throw DeploymentException("Error running jlink: ${err.toString()}" )
        }
    }

    private fun isJdkModule(moduleName: String): Boolean {
        return BackstageConstants.JDK_MODULES_PREFIXES.any { moduleName.startsWith(it.trim()) }

    }

    private fun getAllJarsInDirectory(directory: Path): List<String> {
        return Files.list(directory).filter { it.name.endsWith(BackstageConstants.JAR_EXTENSION) }
            .map { it.toFile().absolutePath }.toList()
    }

    private fun concatenateModulesPath(paths: List<Path>) : String {
        val sb = StringBuilder(paths[0].toAbsolutePath().toString())
        paths.stream().skip(1).forEach { sb.append(":").append(it.toAbsolutePath().toString()) }
        return sb.toString()
    }

    private fun createModulesArgs(modules : Set<String>) : String {
        val sb = StringBuilder()
        modules.forEach { sb.append(",").append(it) }
        return sb.toString().substring(1)
    }
}